### ISSUE TITLE
Don't compile sgame with `-ffast-math`

### ISSUE THREAD

**Issue by necessarily-equal** [Unvanquished/Unvanquished#2311]
`-ffast-math` can cause extra weird reliability issues like https://github.com/Unvanquished/Unvanquished/issues/2299. I suspect we don't need the extra bit of performance on sgame, although it may be useful for cgame and the renderer.

To avoid that kind of issue, how about building sgame without fast-math?

From the gcc doc, here are some problems that can happen with fast-math, beside the one from https://github.com/Unvanquished/Unvanquished/issues/2299#issuecomment-1313325262:

> On 68000 and x86 systems, for instance, you can get paradoxical results if you test the precise values of floating point numbers. For example, you can find that a floating point value which is not a NaN is not equal to itself. This results from the fact that the floating point registers hold a few more bits of precision than fit in a double in memory. Compiled code moves values between memory and floating point registers at its convenience, and moving them into memory truncates them. 
> 
> Do not store floating-point variables in registers, and inhibit other options that might change whether a floating-point value is taken from a register or memory.
> 
> This option prevents undesirable excess precision on machines such as the 68000 where the floating registers (of the 68881) keep more precision than a double is supposed to have. Similarly for the x86 architecture. For most programs, the excess precision does only good, but a few programs rely on the precise definition of IEEE floating point. Use -ffloat-store for such programs, after modifying them to store all pertinent intermediate computations into variables.

Ultimately it's a question of balancing performance and predictable behaviour. I think predictable behavior is more important, at least for sgame.

**Comment by DolceTriade** [comment]
Yeah. Have you investigated any workaround for the code in question or is disabling -ffast-math the only option?

**Comment by illwieckz** [comment]
Quote from https://github.com/Unvanquished/Unvanquished/issues/2299#issuecomment-1313325262 :

> We found out that `-ffloat-store` makes the problem disappear. So this problem would be because of compiler optimisations in floating-point number behavior.

If setting `-ffloat-store` works around the bug, do we need to set the whole `-ffast-math` and potentially disable other optimizations that don't break?

**Comment by necessarily-equal** [comment]
We may decide on whitelisting some optimisations such as -fassociative-math instead of blindly using all of -ffast-math. However, I don't think we'll find the fix soon because I didn't find the affected function, because the bug is quite weird (adding a debug statement can make it disappear) and because for some reason I can't reproduce the bug anymore.

My reasoning for disabling -ffast-math is that :
1. It brings a lot of incertitude, it brings a lot of weird or non-sensical behaviors and it makes troubleshooting stuff a huge pain (because it allows rewrite that can cause or remove bugs depending on the surrounding functions like the bug we are seeing).
2. I don't think there is a large performance impact about disabling fast-math in sgame (sgame doesn't do that much floating point math).
3. We don't know where the bug is really, and we don't want to hunt it more IMO.
4. -ffast-math has a potential of creating more bugs we don't want to start hunting down.

**Comment by slipher** [comment]
float-store and fast-math appear to be orthogonal. If you enable float-store then every time you use a floating point register, it copies the register to the stack and then back into the register. fast-math enables optimizations like converting `x + x + x` to `x * 3`.



> [`-ffast-math`] brings a lot of incertitude, it brings a lot of weird or non-sensical behaviors

Like what? The only example I see is that a value may not be equal to itself, but it is well known that doing an equality test on floats is a bad idea. If you suspect that may be an issue you can check for cases by enable `-Wfloat-equal`.

**Comment by necessarily-equal** [comment]
I've seen more funky things than I thought I would. I think the root of the problem we are experiencing in #2299 is that nan-comparision gives wildly random results:

```C++
check( runtime_nan > 1.0f ); // true
check( runtime_nan == 1.0f ); // true
check( runtime_nan < 1.0f ); // true
check( runtime_nan > 2.0f ); // false
check( runtime_nan < 2.0f ); // true
check( runtime_nan == 2.0f ); // true
check( runtime_nan >= 2.0f ); // false
```

Like, for some reason nan is bigger than 1 but also it is smaller than 1 and equal to 1. But's it's not bigger than 2. Because apparently it's smaller than 2 — but wait, nan is also smaller than 2. But somehow while nan is equal to 2 it's not bigger or equal than 2 !

This is likely why only a part of the map gets nuked in #2299.

Investigating #2299 is made even harder by the optimisations playing with each another. If you add a printf, you can suddenly make it work. But of course you didn't fix the problem! This is a bit like playing with quantum mechanics and not being able to observe the system without interfering.

**Comment by slipher** [comment]
For a Linux DLL build you can supposedly make it crash on float division by 0 by doing this:

```
feenableexcept(FE_DIVBYZERO); 
```

Maybe you can find NaNs this way.

**Comment by odiferousmint** [comment]
You said `NaN`? `-ffast-math`?! Readings for the Sunday:

https://gcc.gnu.org/bugzilla/show_bug.cgi?id=84949

https://lists.llvm.org/pipermail/llvm-dev/2021-September/152530.html

https://pspdfkit.com/blog/2021/understanding-fast-math/

https://news.ycombinator.com/item?id=28927064

https://kristerw.github.io/2021/10/26/fast-math-ub/

https://kristerw.github.io/2021/10/19/fast-math/

**Comment by slipher** [comment]
I've classified some recent bugs believed to be triggered by fast-math mode and classified them in the following system: 
- __0__: our code never had any right to work
- __1__: our code did not follow floating point best practices but it should indeed work under a direct translation of the operations
- __2__: the compiler fucked up our shit without any justification

| Bug | Module | Description | Code
| --- | --- | --- | ---
| https://github.com/Unvanquished/Unvanquished/issues/2299 | sgame | mover with nan/inf velocity destroys everything | 0
| https://github.com/DaemonEngine/Daemon/issues/1518 | engine | light grid size computation broken | 1
| https://github.com/DaemonEngine/Daemon/issues/1527 | engine | models broken via quaternion function | unknown
| https://github.com/Unvanquished/Unvanquished/issues/3140 | cgame | portals broken via quaternion function | unknown
| https://github.com/DaemonEngine/Daemon/issues/1312 | engine | models disappear because clang miscompiles bitwise sse instructions | 2

I'd say we have enough data now to disable fast-math everywhere. At the time the issue was created, we knew only the bug that I classify as a 0.

**Comment by illwieckz** [comment]
I did some comparisons on two machines:

- o740: a low-end machine having an `AMD Athlon 64 X2 Dual Core Processor 4400+ CPU` (Toledo) and an `ATI X550 HyperMemory GPU` (RV370), it was running Unvanquished on a bare Xorg.
- xobx, a middle-end machine having an `AMD 4700S 8-Core Processor Desktop Kit CPU` (Zen 2) and an `AMD Radeon RX 6700 XT` (Navi 22, RDNA 2.0) and some GDDR6 as main CPU memory (this is the actual Xbox Series CPU on a PC motherboard with the retail GPU the closest to the Xbox Series GPU), it was running Unvanquished on a Xorg with GNOME Shell.

The plat23 “alien” scene is `1920 1920 0 0 0`.
The “no VBO” option is `r_vboModel 0`.

A framerate prefixed with`~` means it's not stable and I picked the highest I have seen (that's not really reliable).

All builds are using `RelwithDebInfo` build type and LTO (like our releases). I was running exe game for quick rebuild.
For the “xobx” machine, I also tested with an AVX build.

|machine<br/>resolution<br/>preset|SIMD|fast|plat23<br/>spectator|plat23<br/>alien|plat23<br/>alien<br/>no VBO|thunder<br/>spectator|0.55.2<br/>benchmark<br/>demo
|-|-|-|-|-|-|-|-
|**o740<br/>640×480<br/>lowest**|SSE|✅️ Yes|117 fps|45 fps|26 fps|104 fps|6204 frames, 101.8s: 60.9 fps
|**o740<br/>640×480<br/>lowest**|SSE|❌️ No|117 fps|45 fps|26 fps|104 fps|6204 frames, 102.4s: 60.6 fps
|**xobx<br/>640×480<br/>lowest**|SSE|✅️ Yes|~1899 fps|~1078 fps|73 fps|~1171 fps|6204 frames, 9.2s: 672.8 fps
|**xobx<br/>640×480<br/>lowest**|SSE|❌️ No|~1868 fps|~1095 fps|73 fps|~1991 fps|6204 frames, 8.8s: 705.5 fps
|**xobx<br/>640×480<br/>lowest**|AVX|✅️ Yes|~1843 fps|~1100 fps|87 fps|~1224 fps|6204 frames, 9.1s: 680.0 fps
|**xobx<br/>640×480<br/>lowest**|AVX|❌️ No|~1839 fps|~1062 fps|87 fps|~1225 fps|6204 frames, 9.0s: 692.9 fps
|**xobx<br/>3840×2160<br/>ultra**|SSE|✅️ Yes|383 fps|407 fps|67 fps|329 fps|6204 frames, 16.0s: 388.1 fps
|**xobx<br/>3840×2160<br/>ultra**|SSE|❌️ No|384 fps|408 fps|68 fps|330 fps|6204 frames, 15.9s: 390.1 fps
|**xobx<br/>3840×2160<br/>ultra**|AVX|✅️ Yes|383 fps|407 fps|80 fps|329 fps|6204 frames, 16.1s: 386.4 fps
|**xobx<br/>3840×2160<br/>ultra**|AVX|❌️ No|384 fps|407 fps|80 fps|329 fps|6204 frames, 16.0s: 388.6 fps

I don't see that much of a benefit (euphemism), in fact the SSE ultra 4K on xobx was 1 fps faster (the fps was stable) without fast math on all tests. I have tested all those runs only once so take this as it is.
